File: src/breslin/clientside/move/move.php
var Move = new Class(
{

initialize: function (shapeDynamic)
{
	//shape
    	this.mShape = shapeDynamic;

	/******************************** Process **/
	//processTick StateMachine
        this.mProcessTickStateMachine       = new StateMachine(this);    //setup the state machine

	//processTick states	
	this.mGlobalProcessTickMove  = new GlobalProcessTickMove (this); 
	this.mCatchupProcessTickMove = new CatchupProcessTickMove(this); 
	this.mNormalProcessTickMove  = new NormalProcessTickMove (this); 
        
        this.mProcessTickStateMachine.setCurrentState      (this.mNormalProcessTickMove);
        this.mProcessTickStateMachine.setPreviousState     (this.mNormalProcessTickMove);
        this.mProcessTickStateMachine.setGlobalState       (this.mGlobalProcessTickMove);

	/************************************ Interpolate **/
	//interpolateTick StateMachine
        this.mInterpolateTickStateMachine       = new StateMachine(this);    //setup the state machine

	//interpolate states
	this.mNormalInterpolateTickMove  = new NormalInterpolateTickMove(this); 
       
	//interpolateTick states
        this.mInterpolateTickStateMachine.setCurrentState      (this.mNormalInterpolateTickMove);
        this.mInterpolateTickStateMachine.setPreviousState     (this.mNormalInterpolateTickMove);
        this.mInterpolateTickStateMachine.setGlobalState       (0);
   
	/***************************************** variables **/
	//thresholds
    	this.mPosInterpLimitHigh = parseFloat(.066); //how far away from server till we try to catch up
    	this.mPosInterpFactor    = parseFloat(4.0);
        this.mMaximunVelocity    = parseFloat(.003083); //do not let velocity go above this in any direction.

        //deltas
        this.mDeltaX        = 0.0;
        this.mDeltaY        = 0.0;
        this.mDeltaZ        = 0.0;
        this.mDeltaPosition = 0.0;
},

log: function(msg)
{
        setTimeout(function()
        {
                throw new Error(msg);
        }, 0);
},

/******************************************************
*                               UPDATING
********************************************************/
processTick: function()
{
        this.mProcessTickStateMachine.update();
},

interpolateTick: function(renderTime)
{
        this.mInterpolateTickStateMachine.update();
},

/******************************************************
*                               MOVE
********************************************************/
calculateDeltaPosition: function()
{
        this.mDeltaX = this.mShape.mServerCommandCurrent.mPosition.x - this.mShape.getPosition().x;
        this.mDeltaY = this.mShape.mServerCommandCurrent.mPosition.y - this.mShape.getPosition().y;
        this.mDeltaZ = this.mShape.mServerCommandCurrent.mPosition.z - this.mShape.getPosition().z;

        //distance we are off from server
        this.mDeltaPosition = Math.sqrt(Math.pow(this.mDeltaX, 2) + Math.pow(this.mDeltaY, 2) +  Math.pow(this.mDeltaZ, 2));
},

calculateSpeed: function(velocity,frameTime)
{
        speed = Math.sqrt(
        Math.pow(parseFloat(velocity.x), 2) +
        Math.pow(parseFloat(velocity.z), 2)) /
        parseFloat(frameTime);
	
	

        return parseFloat(speed);
}

});

File: src/breslin/clientside/move/move.cpp
#include "move.h"

//log
#include "../tdreamsock/dreamSockLog.h"

//shape
#include "../shape/shape.h"

//command
#include "../command/command.h"

//states
#include "states/moveStates.h"


//std lib
#include <math.h>

Move::Move(Shape* shapeDynamic) 
{
	mShape = shapeDynamic;

	//move processTick states
	mProcessTickStateMachine = new StateMachine<Move>(this);    //setup the state machine
	mProcessTickStateMachine->setCurrentState      (NORMAL_PROCESSTICK_MOVE::Instance());
	mProcessTickStateMachine->setPreviousState     (NORMAL_PROCESSTICK_MOVE::Instance());
	mProcessTickStateMachine->setGlobalState       (GLOBAL_PROCESSTICK_MOVE::Instance());

	//move interpolateTick states
	mInterpolateTickStateMachine = new StateMachine<Move>(this);    //setup the state machine
	mInterpolateTickStateMachine->setCurrentState      (NORMAL_INTERPOLATETICK_MOVE::Instance());
	mInterpolateTickStateMachine->setPreviousState     (NORMAL_INTERPOLATETICK_MOVE::Instance());
	//mInterpolateTickStateMachine->setGlobalState     (GLOBAL_INTERPOLATETICK_MOVE::Instance());
	mInterpolateTickStateMachine->setGlobalState       (NULL);

    	//thresholds
    	mPosInterpLimitHigh = .066f; //how far away from server till we try to catch up
    	mPosInterpFactor    = 4.0f;
	mMaximunVelocity    = .009083f; //do not let velocity go above this in any direction.

	//deltas
	mDeltaX        = 0.0f; 
	mDeltaY	       = 0.0f;
	mDeltaZ        = 0.0f;
	mDeltaPosition = 0.0f;
}

Move::~Move()
{
}

/******************************************************
*				UPDATING
********************************************************/
void Move::processTick()
{
	mProcessTickStateMachine->update();
}
void Move::interpolateTick(float renderTime)
{
	mInterpolateTickStateMachine->update();
}

/******************************************************
*				MOVE
********************************************************/

void Move::calculateDeltaPosition()  //mov
{

	mDeltaX = mShape->mServerCommandCurrent->mPosition->x - mShape->getPosition()->x;
    	mDeltaY = mShape->mServerCommandCurrent->mPosition->y - mShape->getPosition()->y;
    	mDeltaZ = mShape->mServerCommandCurrent->mPosition->z - mShape->getPosition()->z;

    	//distance we are off from server
    	mDeltaPosition = sqrt(pow(mDeltaX, 2) + pow(mDeltaY, 2) +  pow(mDeltaZ, 2));
}

float Move::calculateSpeed(Vector3D* velocity, int frameTime)
{
	float speed = sqrt(
	pow(velocity->x, 2) + 
    	pow(velocity->y, 2) +
	pow(velocity->z, 2)) /
	frameTime;

	return speed;
}

void Move::regulate(Vector3D* velocityToRegulate)
{
	//x
	if (velocityToRegulate->x > mMaximunVelocity)
	{
		velocityToRegulate->x = mMaximunVelocity;
		LogString("x>");

	}
	
	if (velocityToRegulate->x < mMaximunVelocity * -1)
	{
		velocityToRegulate->x = mMaximunVelocity * -1;
		LogString("x<");
	}
	

	//z
	if (velocityToRegulate->z > mMaximunVelocity)
	{
		velocityToRegulate->z = mMaximunVelocity;
		LogString("z>");
	}

	if (velocityToRegulate->z < mMaximunVelocity * -1)
	{
		velocityToRegulate->z = mMaximunVelocity * -1;
		LogString("z<");
	}
}
